src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,810 @@
+{
+    var bob = new Employee { Name = "Bob" };
+    var angela = new Employee { Name = "Angela" };
+
+    angela.Manager = bob;
+    bob.Subordinates = new List<Employee>{ angela };
+    
+    var allEmployees = new List<Employee>
+    {
+        angela,
+        bob
+    };
+
+    var options = new JsonSerializerOptions
+    {
+        ReferenceHandling = new ReferenceHandling(
+            PreserveReferencesHandling.All, // Preserve References Handling on serialization.
+            PreserveReferencesHandling.All, // Preserve References Handling on deserialization.
+            ReferenceLoopHandling.Ignore) // Reference Loop Handling on serialization.
+        WriteIndented = true,
+    };
+
+    string json = JsonSerializer.Serialize(allEmployees, options);
+    Console.Write(json);
+
+    /* Output:
+    [
+        {
+            "$id": "1",
+            "Name": "Angela",
+            "Manager": {
+                "$id": "2",
+                "Name": "Bob",
+                "Subordinates": {
+                    "$id": "3",
+                    // Note how subordinates is empty due Angela is being ignored.
+                    // Alternatively: we may let PreserveReferenceHandling take precedence and write the reference instead?
+                    "$values": []
+                }
+            }
+        },
+        {
+            // Note how element 2 is written as a reference 
+            // since was previously seen in allEmployees[0].Manager
+            "$ref": "2"
+        }
+    ]
+    */
+
+    allEmployees = JsonSerializer.Deserialize<List<Employee>>(json, options);
+    Console.WriteLine(allEmployees[0].Manager == allEmployees[1]); 
+    /* Output: true */
+}
+```
+
+# Notes
+
+1. MaxDepth validation will not be affected by `ReferenceHandling.Preserve`.
+2. We are merging the Newtonsoft.Json types [`ReferenceLoopHandling`]("https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_ReferenceLoopHandling.htm"), [`MetadataPropertyHandling`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_MetadataPropertyHandling.htm) (without `ReadAhead`), and [`PreserveReferencesHandling`]("https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_PreserveReferencesHandling.htm") (without the granularity of `Objects` and `Arrays`) into one single class; `ReferenceHandling`.
+3. While Immutable types and `System.Array`s can be Serialized with Preserve semantics, they will not be supported when trying to Deserialize them as a reference; those types are created with the help of an internal converter, and they are not parsed until the entire block of JSON finishes, nested reference to these types is impossible to identify, unless you re-scan the resulting object, which is too expensive.
+4. Value types, such as structs that contain preserve semantics, will not be supported when Deserialized as well; this is because the serializer will never signal a reference object to those types, doing such thing implies boxing of value types.
```

Comment:
```suggestion
4. Value types, such as structs that contain preserve semantics, will not be supported when deserializing as well. This is because the serializer will never emit a reference object to those types and doing so implies boxing of value types.
```